AK/Optional.h

```diff
@@ -632,6 +632,171 @@ requires(IsLvalueReference<T>) class [[nodiscard]] Optional<T> {
+            value() = other.release_value();
+        } else if (has_value()) {
+            value().~T();
+            m_value = the_empty_value();
+        } else if (other.has_value()) {
+            m_value = other.release_value();
+        }
+        return *this;
+    }
+
+    ALWAYS_INLINE constexpr Optional(Optional&& other)
+    requires(dont_move_empty && has_swap)
+    {
+        swap(m_value, other.m_value);
+        other.clear();
+    }
+    ALWAYS_INLINE constexpr Optional& operator=(Optional&& other)
+    requires(dont_move_empty && has_swap)
+    {
+        if (this == &other)
+            return *this;
+        swap(m_value, other.m_value);
+        other.clear();
+        return *this;
+    }
+
+    ALWAYS_INLINE constexpr bool operator==(Optional const& other) const
+    {
+        return has_value() == other.has_value() && (!has_value() || value() == other.value());
+    }
+
+    template<typename U>
+    requires(
+        !IsScalar<U> && IsConstructible<T, U const&>
+        && !IsOneOf<RemoveCVReference<U>, Optional, OptionalNone>)
+    ALWAYS_INLINE constexpr Optional& operator=(U const& value)
+    {
+        m_value = value;
+        return *this;
+    }
+    template<typename U>
+    requires(
+        !IsScalar<U> && IsConstructible<T, U &&>
+        && !IsOneOf<RemoveCVReference<U>, Optional, OptionalNone>)
+    ALWAYS_INLINE constexpr Optional& operator=(U&& value)
+    {
+        m_value = forward<U>(value);
+        return *this;
+    }
+
+    ALWAYS_INLINE constexpr bool has_value() const
+    {
+        return m_value != the_empty_value();
+    }
+    ALWAYS_INLINE constexpr bool has_value() const
+    requires(requires(T const& t) {{ Traits<T>::optional_has_value(t) } -> SameAs<bool>; })
+    {
+        return Traits<T>::optional_has_value(m_value);
+    }
+    ALWAYS_INLINE constexpr void clear()
+    {
+        if (has_value())
+            (void)release_value();
+    }
+
+    ALWAYS_INLINE constexpr T& value() &
+    {
+        VERIFY(has_value());
+        return m_value;
+    }
+    ALWAYS_INLINE constexpr T const& value() const&
+    {
+        VERIFY(has_value());
+        return m_value;
+    }
+
+    ALWAYS_INLINE constexpr T value() && { return release_value(); }
+
+    ALWAYS_INLINE constexpr T release_value()
+    {
+        // FIXME: What to do here when we should not move empties around,
+        //        But don't have a swap method to place it in?
+        VERIFY(has_value());
+        return exchange(m_value, the_empty_value());
+    }
+
+    ALWAYS_INLINE constexpr T release_value()
+    requires(dont_move_empty && has_swap)
+    {
+        VERIFY(has_value());
+        T value = the_empty_value();
+        value.swap(m_value);
+        return value;
+    }
+
+    template<typename... Ts>
+    requires(IsConstructible<T, Ts...>)
+    ALWAYS_INLINE constexpr void emplace(Ts&&... parameters)
+    {
+        clear();
+        m_value = T(forward<Ts>(parameters)...);
```

Comment:
not sure if `construct_at` is nicer here as `m_value` is usually in a replaceable state here